src/Definition.elm

```diff
@@ -140,12 +140,22 @@ decodeTypes =
 
 decodeTermDefInfo : Decode.Decoder TermDefinitionInfo
 decodeTermDefInfo =
+    let
+        decodeTermDefTag =
+            at [ "termDefinition", "tag" ] Decode.string
+
+        decodeUserObject =
+            Decode.map TermSource (at [ "termDefinition", "contents" ] Syntax.decode)
+
+        decodeBuiltin =
+            Decode.map (TypeSignature >> BuiltinTerm) (field "signature" Syntax.decode)
+    in
     Decode.map3 TermDefinitionInfo
         (field "termNames" (Util.decodeNonEmptyList FullyQualifiedName.decode))
         (field "bestTermName" Decode.string)
         (Decode.oneOf
-            [ Decode.map TermSource (at [ "termDefinition", "contents" ] Syntax.decode)
-            , Decode.map (TypeSignature >> BuiltinTerm) (field "signature" Syntax.decode)
+            [ when decodeTermDefTag ((==) "UserObject") decodeUserObject
+            , when decodeTermDefTag ((==) "BuiltinObject") decodeBuiltin
```

Comment:
Possibly a bit cleaner - I'm guessing you could implement a combinator like:

```Elm
Decoder.oneOfTagged : [(String, Decoder a)] -> Decoder a
```

Which encapsulates that dispatching on `tag` and applying the matching decoder to `contents`, and failing with an "unknown tag" if there's no matching decoder.

Reply:
Nice, I like that!